File: frontend/js/controller/Controller.js
export class Controller {

    static #instance = null;
    static #API_URL = "/api/requests";

    constructor() {
        throw new Error("Используйте Controller.getInstance() вместо new");
    }

    static getInstance() {
        if (!Controller.#instance) {
            Controller.#instance = Object.create(Controller.prototype);
        }
        return Controller.#instance;
    }

    async getStudents(filters = {}) {
        const url = new URL(this.#API_URL);

        const allowedFilters = [
            "isuId",
            "fio",
            "stGroup",
            "dormitoryNumber",
            "room",
            "dateOfPlacement",
            "isNotRussian",
        ];

        for (const key of allowedFilters) {
            const value = filters[key];

            if (value !== undefined && value !== null && value !== "") {
                url.searchParams.set(key, String(value));
            }
        }

        return await this.request_response_cycle(url.toString(), {
            method: "GET"
        }
        );
    }

    async getStudent(id) {
        return await this.request_response_cycle(
            `${this.#API_URL}/${encodeURIComponent(id)}`, {
            method: "GET"
        }
        );
    }

    async addStudent(student) {
        return await this.request_response_cycle(this.#API_URL, {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify(student)
        }
        );
    }

    async updateStudent(id, student) {
        return await this.request_response_cycle(
            `${this.#API_URL}/${encodeURIComponent(id)}`, {
            method: "PATCH",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify(student)
        }
        );
    }

    async deleteStudent(id) {
        return await this.request_response_cycle(
            `${this.#API_URL}/${encodeURIComponent(id)}`, {
            method: "DELETE"
        }
        );
    }

    async request_response_cycle(url, options = {}) {
        let response;
        try {
            response = await fetch(url, {
                ...options,
                headers: {
                    Accept: "application/json",
                    ...options.headers
                }
            });
        } catch (error) {
            throw this.createNetworkError(error);
        }
        return await this.processResponse(response);
    }

    async processResponse(response) {
        if (response.status === 204) {
            return null;
        }

        const data = await this.parseResponseBody(response);

        if (!response.ok) {
            const error = new Error(
                data?.error?.message ||
                data?.message ||
                `HTTP error: ${response.status}`
            );

            error.status = response.status;
            error.code = data?.error?.code;

            throw error;
        }
        return data;
    }

    async parseResponseBody(response) {
        const contentType = response.headers.get("content-type") || "";

        if (contentType.includes("application/json")) {
            try {
                return await response.json();
            } catch {
                return null;
            }
        }
        return null;
    }

    createNetworkError(error) {
        if (error?.status !== undefined) {
            return error;
        }

        const networkError = new Error(
            "Не удалось связаться с сервером"
        );

        networkError.code = "NETWORK_ERROR";
        networkError.originalError = error;

        return networkError;
    }
}

File: backend/controller/controller.py
from flask import Blueprint, jsonify, request

from backend.exceptions import ServerException, NotFoundException, ValidationException, NotUniqueIdException

students_bp = Blueprint("students",
                        __name__,
                        url_prefix="/students"
                        )

service = None

@students_bp.get("")
def get_students():
    try:
        filters = {
            "isuId": request.args.get("isuId"),
            "fio": request.args.get("fio"),
            "stGroup": request.args.get("stGroup"),
            "dormitoryNumber": request.args.get("dormitoryNumber"),
            "room": request.args.get("room"),
            "dateOfPlacement": request.args.get("dateOfPlacement"),
            "isNotRussian": request.args.get("isNotRussian")
        }

        students = service.get_students(filters)
        return jsonify(students), 200
    except ServerException:
        return jsonify({
            "error": {
                "code": "INTERNAL_SERVER_ERROR",
                "message": "The server cannot process the error"
            }
        }), 500

@students_bp.route("/<int:student_id>", methods=["GET", "QUERY"])
def get_student(student_id):
    try:
        if request.method == "QUERY":
            if not request.is_json:
                return jsonify({
                    "error": {
                        "code": "BAD_REQUEST",
                        "message": "QUERY request body must be JSON"
                    }
                }), 400

            query_data = request.get_json()
            student = service.query_student(student_id, query_data)
            return jsonify(student), 200

        student = service.get_student(student_id)
        return jsonify(student), 200

    except NotFoundException:
        return jsonify({
            "error": {
                "code": "NOT_FOUND",
                "message": f"Student by ISU_ID={student_id} not found"
            }
        }), 404
    except ServerException:
        return jsonify({
            "error": {
                "code": "INTERNAL_SERVER_ERROR",
                "message": "The server cannot process the error"
            }
        }), 500

@students_bp.post("")
def create_student():
    if not request.is_json:
        return jsonify({
            "error": {
                "code": "BAD_REQUEST",
                "message": "Request body must be JSON"
            }
        }), 400

    data = request.get_json()

    try:
        student = service.create_student(data)
        return jsonify(student), 201
    except ValidationException as error:
        return jsonify({
            "error": {
                "code": "VALIDATION_ERROR",
                "message": str(error)
            }
        }), 422
    except NotUniqueIdException:
        return jsonify({
            "error": {
                "code": "CONFLICT",
                "message": "Student with this ISU_ID already exists"
            }
        }), 409
    except ServerException:
        return jsonify({
            "error": {
                "code": "INTERNAL_SERVER_ERROR",
                "message": "The server cannot process the error"
            }
        }), 500

@students_bp.patch("/<int:student_id>")
def update_student(student_id):
    if not request.is_json:
        return jsonify({
            "error": {
                "code": "BAD_REQUEST",
                "message": "Request body must be JSON"
            }
        }), 400

    data = request.get_json()

    try:
        student = service.update_student(student_id, data)
        return jsonify(student), 200
    except ValidationException as error:
        return jsonify({
            "error": {
                "code": "VALIDATION_ERROR",
                "message": str(error)
            }
        }), 422
    except NotFoundException:
        return jsonify({
            "error": {
                "code": "NOT_FOUND",
                "message": f"Student by ISU_ID={student_id} not found"
            }
        }), 404
    except ServerException:
        return jsonify({
            "error": {
                "code": "INTERNAL_SERVER_ERROR",
                "message": "The server cannot process the error"
            }
        }), 500

@students_bp.delete("/<int:student_id>")
def delete_student(student_id):
    try:
        service.delete_student(student_id)
        return "", 204
    except ValidationException as error:
        return jsonify({
            "error": {
                "code": "VALIDATION_ERROR",
                "message": str(error)
            }
        }), 422
    except NotFoundException:
        return jsonify({
            "error": {
                "code": "NOT_FOUND",
                "message": f"Student by ISU_ID={student_id} not found"
            }
        }), 404
    except ServerException:
        return jsonify({
            "error": {
                "code": "INTERNAL_SERVER_ERROR",
                "message": "The server cannot process the error"
            }
        }), 500
